crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1107,7 +1107,27 @@ impl OlmMachine {
             self.store.get_inbound_group_session(room_id, content.session_id()).await?
         {
             // TODO check the message index.
-            let (decrypted_event, _) = session.decrypt(event).await?;
+            let decrypt_result = session.decrypt(event).await;
+
+            if let Err(decrypt_error) = decrypt_result {
+                if let MegolmError::Decryption(ref vodo_error) = decrypt_error {
+                    if let DecryptionError::UnknownMessageIndex(known_index,message_index) = vodo_error {
+
+                        trace!(
+                            "Unabled to decrypt {}, ratched key {}, message index {}",
+                            session.session_id(),
+                            known_index,
+                            message_index
+                        );
+
+                        // It's a ratcheted key, worth sending a request to our other sessions
+                        self.key_request_machine.create_outgoing_key_request(room_id, event).await?;
+                    }
+                }
+                return Err(decrypt_error);
+            } 
+
+            let (decrypted_event, _) = decrypt_result.unwrap();
```

Comment:
This is general code smell, I know that this won't panic because there's a return just above this, but some small restructuring or new feature work might remove that guarantee.

Early returns + unwraps quickly become unmanageable and hard to reason about things. My initial suggestion of moving the requesting into the caller of this method will get rid of this problem.

Reply:
👍